index.js

```diff
@@ -122,36 +127,52 @@ async function openTelemetryPlugin (fastify, opts = {}) {
   }
 
   async function onError (request, reply, error) {
-    if (shouldIgnoreRoute(request.url, request.method)) return
-
     const activeContext = getContext(request)
     const span = trace.getSpan(activeContext)
     span.setAttributes(formatSpanAttributes.error(error))
   }
 
-  fastify.addHook('onRequest', onRequest)
-  fastify.addHook('onResponse', onResponse)
-  fastify.addHook('onError', onError)
-
-  if (wrapRoutes) {
-    function wrapRoute (routeHandler) {
-      return async function (request, ...args) {
-        const reqContext = getContext(request)
-        return context.with(reqContext, routeHandler.bind(this, request, ...args))
+  function executeIfNotWrappedOrNotIgnoredRoute (handler) {
+    return async function (request, ...params) {
+      if (!shouldWrapRoute(request.routerPath) && !shouldIgnoreRoute(request.url, request.method)) {
+        return handler(request, ...params)
       }
     }
+  }
+
+  fastify.addHook('onRequest', onRequest)
+  fastify.addHook('onResponse', executeIfNotWrappedOrNotIgnoredRoute(onResponse))
+  fastify.addHook('onError', executeIfNotWrappedOrNotIgnoredRoute(onError))
 
-    fastify.addHook('onRoute', function (routeOpts) {
-      const { path, handler, method } = routeOpts
-      if (!shouldIgnoreRoute(path, method)) {
-        if (wrapRoutes === true) {
-          routeOpts.handler = wrapRoute(handler)
-        } else if (Array.isArray(wrapRoutes) && wrapRoutes.includes(path)) {
-          routeOpts.handler = wrapRoute(handler)
+  fastify.addHook('onRoute', function (routeOpts) {
+    const { path, handler, method } = routeOpts
+
+    if (!shouldIgnoreRoute(path, method) && shouldWrapRoute(path)) {
+      function wrapHandler (handler) {
+        return async function (request, ...args) {
+          const reqContext = getContext(request)
+          return context.with(reqContext, handler.bind(this, request, ...args))
         }
       }
-    })
-  }
+
+      routeOpts.handler = wrapHandler(handler)
+
+      function normalizeHooks (hooks) {
```

Comment:
Could we update this function to wrap each hook in `context`? Similar to your suggestion -> https://github.com/autotelic/fastify-opentelemetry/issues/56#issuecomment-1444841874

```js
function wrapHooks (hooks) {
  if (!hooks) {
    return []
  }

  if (Array.isArray(hooks)) {
    return hooks.map(wrapHandler)
  }

  return [wrapHandler(hooks)]
}

// Then just update remaining hook types...
routeOpts.onRequest = [onRequest, ...wrapHooks(routeOpts.onRequest)]
routeOpts.preParsing = wrapHooks(routeOpts.preParsing)
routeOpts.preValidation = wrapHooks(routeOpts.preValidation)
// ... etc.
```